streams/src/test/java/org/apache/kafka/streams/integration/PauseResumeIntegrationTest.java

```diff
@@ -0,0 +1,440 @@
+        return properties;
+    }
+
+    @After
+    public void shutdown() throws InterruptedException {
+        if (kafkaStreams != null) {
+            kafkaStreams.close(Duration.ofSeconds(30));
+        }
+        CLUSTER.deleteTopicsAndWait(INPUT_STREAM_1);
+        CLUSTER.deleteTopicsAndWait(INPUT_STREAM_2);
+        CLUSTER.deleteTopicsAndWait(OUTPUT_STREAM_1);
+        CLUSTER.deleteTopicsAndWait(OUTPUT_STREAM_2);
+    }
+
+    private static void produceToInputTopics(final String topic,
+        final Collection<KeyValue<String, Long>> records) {
+        IntegrationTestUtils.produceKeyValuesSynchronously(
+            topic,
+            records,
+            producerConfig,
+            CLUSTER.time
+        );
+    }
+
+
+    @Test
+    public void shouldPauseAndResumeKafkaStreams() throws Exception {
+        // Create KafkaStreams instance
+        final StreamsBuilder builder = new StreamsBuilder();
+        builder.stream(INPUT_STREAM_1).groupByKey().count().toStream().to(OUTPUT_STREAM_1);
+
+        kafkaStreams = new KafkaStreams(builder.build(props()), props());
+        kafkaStreams.start();
+        waitForApplicationState(singletonList(kafkaStreams), State.RUNNING, STARTUP_TIMEOUT);
+
+        // Write data
+        produceToInputTopics(INPUT_STREAM_1, STANDARD_INPUT_DATA);
+
+        // Verify output
+        assertThat(waitUntilMinKeyValueRecordsReceived(consumerConfig, OUTPUT_STREAM_1, 3),
+            CoreMatchers.equalTo(COUNT_OUTPUT_DATA));
+
+        // Pause KafkaStreams/topology
+        kafkaStreams.pause();
+        assertTrue(kafkaStreams.isPaused());
+
+        // Write more data
+        produceToInputTopics(INPUT_STREAM_1, STANDARD_INPUT_DATA);
+
+        // Verify that consumers read new data -- AKA, there is no lag.
+        final Map<String, Map<Integer, LagInfo>> lagMap = kafkaStreams.allLocalStorePartitionLags();
+        assertNoLag(lagMap);
+
+        // Verify no output somehow?
+        // Is there a better way to show this?
+        assertThat(waitUntilMinKeyValueRecordsReceived(consumerConfig, OUTPUT_STREAM_1, 0),
+            CoreMatchers.equalTo(Collections.emptyList()));
+
+        // Resume KafkaStreams/topology
+        kafkaStreams.resume();
+        assertFalse(kafkaStreams.isPaused());
+
+        // Verify that the input written during the pause is processed.
+        assertThat(waitUntilMinKeyValueRecordsReceived(consumerConfig, OUTPUT_STREAM_1, 3),
+            CoreMatchers.equalTo(COUNT_OUTPUT_DATA2));
+    }
+
+    @Test
+    public void shouldAllowForTopologiesToStartPaused() throws Exception {
+        // Create KafkaStreams instance
+        final StreamsBuilder builder = new StreamsBuilder();
+        builder.stream(INPUT_STREAM_1).groupByKey().count().toStream().to(OUTPUT_STREAM_1);
+
+        kafkaStreams = new KafkaStreams(builder.build(props()), props());
+
+        // Start KafkaStream with paused processing.
+        kafkaStreams.pause();
+        kafkaStreams.start();
```

Comment:
I think the names here are a little confusing.  With context, it makes sense that this is how to start a `KafkaStreams` instance with processing paused.

If anyone has a naming suggestion here, I'm very open to it!